internal/controllers/core/cmd/controller_old.go

```diff
@@ -9,183 +9,174 @@ import (
+	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/tilt-dev/tilt/internal/store"
-	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
 	"github.com/tilt-dev/tilt/pkg/logger"
 	"github.com/tilt-dev/tilt/pkg/model"
 	"github.com/tilt-dev/tilt/pkg/model/logstore"
 )
 
-type ControllerOld struct {
+type Controller struct {
+	client.Client
+
+	st            store.RStore
 	execer        Execer
-	procs         map[model.ManifestName]*currentProcess
+	procs         map[types.NamespacedName]*currentProcess
 	procCount     int
 	proberManager ProberManager
+	mu            sync.Mutex
 }
 
-var _ store.Subscriber = &ControllerOld{}
-var _ store.TearDowner = &ControllerOld{}
-
-func NewControllerOld(execer Execer, proberManager ProberManager) *ControllerOld {
-	return &ControllerOld{
+func NewController(st store.RStore, execer Execer, proberManager ProberManager) *Controller {
+	return &Controller{
+		st:            st,
 		execer:        execer,
-		procs:         make(map[model.ManifestName]*currentProcess),
 		proberManager: proberManager,
+		procs:         make(map[types.NamespacedName]*currentProcess),
 	}
 }
 
-func (c *ControllerOld) OnChange(ctx context.Context, st store.RStore) {
-	specs := c.determineServeSpecs(ctx, st)
-	c.update(ctx, specs, st)
+func (c *Controller) SetClient(client client.Client) {
+	c.Client = client
 }
 
-func (c *ControllerOld) determineServeSpecs(ctx context.Context, st store.RStore) []ServeSpec {
-	state := st.RLockState()
-	defer st.RUnlockState()
-
-	var r []ServeSpec
-
-	for _, mt := range state.Targets() {
-		if !mt.Manifest.IsLocal() {
-			continue
-		}
-		lt := mt.Manifest.LocalTarget()
-		if lt.ServeCmd.Empty() ||
-			mt.State.LastSuccessfulDeployTime.IsZero() {
-			continue
+func (c *Controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var cmd Cmd
+	err := c.Get(ctx, req.NamespacedName, &cmd)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			c.stop(req.NamespacedName)
 		}
-		r = append(r, ServeSpec{
-			ManifestName:   mt.Manifest.Name,
-			ServeCmd:       lt.ServeCmd,
-			TriggerTime:    mt.State.LastSuccessfulDeployTime,
-			ReadinessProbe: lt.ReadinessProbe,
-		})
+		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	return r
-}
-
-func (c *ControllerOld) update(ctx context.Context, specs []ServeSpec, st store.RStore) {
-	var toStop []model.ManifestName
-	var toStart []ServeSpec
-
-	seen := make(map[model.ManifestName]bool)
-	for _, spec := range specs {
-		seen[spec.ManifestName] = true
-		proc := c.getOrMakeProc(spec.ManifestName)
-		if c.shouldStart(spec, proc) {
-			toStart = append(toStart, spec)
-		}
+	if !cmd.ObjectMeta.DeletionTimestamp.IsZero() {
+		c.stop(req.NamespacedName)
+		return ctrl.Result{}, nil
 	}
 
-	for name := range c.procs {
-		if !seen[name] {
-			toStop = append(toStop, name)
-		}
-	}
-
-	// stop old ones
-	for _, name := range toStop {
-		c.stop(name)
-		delete(c.procs, name)
-	}
+	c.ensureStarted(ctx, req.NamespacedName, &cmd)
```

Comment:
You'll need to create a new context from `context.Background()` so that processes don't get killed as soon as `Reconcile()` completes

Reply:
i was worried about this too, but I don't think it's true! at least, when I tested it, i didn't see this behavior.

Reply:
I just dug through the source and you're right - guess my context was getting canceled by something else at the time 🙃 

FWIW The context arg is new in `controller-runtime` v0.8.0, which kubebuilder docs haven't actually been updated for yet. The godoc doesn't actually say anything specific about the context/lifetime either.